_overviews/scala3-book/domain-modeling-fp.md

```diff
@@ -0,0 +1,408 @@
+
+````
+0, 1, 2 ... Int.MaxInt
+````
+
+Ignoring the division of whole numbers, the possible *operators* on those values are:
+
+````
++, -, *
+````
+
+An FP design is implemented in a similar way:
+
+- You have a set of values
+- You have a collection of operators that work on those values
+
+In this chapter we’ll model the data and operations for a “pizza” in a pizza store. You’ll see how to implement the “data” portion of the Scala/FP model, and then you’ll see several different ways you can organize the operations on that data.
+
+
+
+## Modeling the data
+
+Modeling the “data” portion of a domain model in Scala is simple:
+
+- Create ADTs and GADTs with the `enum` construct
+- Use `case` classes for more complicated systems
+<!-- TODO: wording on that last line? -->
+
+
+### Crust size, crust type, and toppings
+
+Data concepts like crust size, crust type, and toppings are concisely modeled with the Scala 3 `enum` construct:
+
+```scala
+enum CrustSize:
+  case Small, Medium, Large
+
+enum CrustType:
+  case Thin, Thick, Regular
+
+enum Topping:
+  case Cheese, Pepperoni, BlackOlives, GreenOlives, Onions
+```
+
+
+### Modeling a pizza
+
+Now you can model a `Pizza` class using a `case` class:
+
+```scala
+import CrustSize._
+import CrustType._
+import Topping._
+
+case class Pizza(
+  crustSize: CrustSize,
+  crustType: CrustType,
+  toppings: Seq[Topping]
+)
+```
+
+And that’s it. That’s the data model for an FP-style pizza system. This solution is very concise because it doesn’t require the operations on a pizza to be combined with the data model. The data model is easy to read, like declaring the design for a relational database.
+
+<!-- TODO: compare this with C-style structs and others? -->
+
+#### More of the data model
+
+Because that was very short, you may prefer to see more of the data model for an entire pizza-ordering system. Here are a few other `case` classes that are used to model such a system:
+
+```scala
+case class Address(
+  street1: String,
+  street2: Option[String],
+  city: String,
+  state: String,
+  zipCode: String
+)
+
+case class Customer(
+  name: String,
+  phone: String,
+  address: Address
+)
+
+case class Order(
+  pizzas: Seq[Pizza],
+  customer: Customer
+)
+```
+
+#### “Skinny domain objects”
+
+In his book, *Functional and Reactive Domain Modeling*, Debasish Ghosh states that where OOP practitioners describe their classes as “rich domain models” that encapsulate data and behaviors, FP data models can be thought of as “skinny domain objects.” This is because — as this lesson shows — the data models are defined as `case` classes with attributes, but no behaviors.
+
+
+
+
+## What about the operations?
+
+This leads to an interesting question: Because FP separates the data from the operations on that data, how do you implement those operations in Scala/FP?
```

Comment:
```suggestion
This leads to an interesting question: Because FP separates the data from the operations on that data, how do you implement those operations in Scala?
```